launch_ros/launch_ros/utilities/evaluate_parameters.py

```diff
@@ -52,12 +54,29 @@ def evaluate_parameter_dict(
             if isinstance(value[0], Substitution):
                 # Value is a list of substitutions, so perform them to make a string
                 evaluated_value = perform_substitutions(context, list(value))
+                evaluated_value = yaml.safe_load(evaluated_value)
             elif isinstance(value[0], Sequence):
                 # Value is an array of a list of substitutions
                 output_subvalue = []  # List[str]
                 for subvalue in value:
-                    output_subvalue.append(perform_substitutions(context, list(subvalue)))
+                    value = perform_substitutions(context, list(subvalue))
+                    output_subvalue.append(value)
                     evaluated_value = tuple(output_subvalue)
+                # All values in a list must have the same type.
+                # If they don't then assume it is a list of strings
+                yaml_evaluated_value = []  # Union[List[str], List[int], List[float], List[bool]]
+                type_ = None
+                dissimilar_types = False
+                for subvalue in evaluated_value:
+                    yaml_subvalue = yaml.safe_load(subvalue)
+                    subtype = type(yaml_subvalue)
+                    if type_ is not None and type_ != subtype:
```

Comment:
When is `type_` `None`?

Reply:
`type_` is `None` when looking at the first element of the list. A better name would be `last_subtype`

Reply:
`type_` -> `last_subtype` in 032ae59